rust/src/ssh/detect.rs

```diff
@@ -0,0 +1,108 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use crate::core::{STREAM_TOCLIENT, STREAM_TOSERVER};
+use super::ssh::SSHTransaction;
+use std::ptr;
+
+#[no_mangle]
+pub extern "C" fn rs_ssh_tx_get_protocol(
+    tx: *mut std::os::raw::c_void,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+    direction: u8,
+) -> u8 {
+    let tx = cast_pointer!(tx, SSHTransaction);
+    match direction {
+        STREAM_TOSERVER => {
+            let m = &tx.cli_hdr.protover;
+            if m.len() > 0 {
+                unsafe { *buffer = m.as_ptr() };
```

Comment:
Can we do less `unsafe` in these functions ?
Like checking if the pointer is not NULL

Reply:
You can at least wrap both `*buffer` and `*buffer_len` assignments in a single `unsafe`. We need `unsafe` when we pass something to C though.

Reply:
My understanding is that we need `unsafe` here because we use a pointer _from_ C
So, can we do anything about it, or just wrap the two statements together ?

Reply:
The from C part is hidden by the `cast_pointer` macro. The `unsafe` here is for returning pointer and length value to the caller. This also crosses the FFI boundary, and is thus unsafe.

Reply:
In this case its probably best to just mark the whole function as unsafe.  Clippy  would prefer this as well:
```
12465 error: this public function dereferences a raw pointer but is not marked `unsafe`
12466   --> src/ssh/detect.rs:34:27
12467    |
12468 34 |                 unsafe { *buffer = m.as_ptr() };
12469    |                           ^^^^^^
12470    |
12471    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#not_unsafe_ptr_arg_deref
```

Reply:
My compiler produces a warning when the function is marked unsafe :
```
warning: unnecessary `unsafe` block
   --> src/parser.rs:127:31
    |
127 |       ($ptr:ident, $ty:ty) => ( unsafe{ &mut *($ptr as *mut $ty) } );
    |                                 ^^^^^^ unnecessary `unsafe` block
    | 
   ::: src/ssh/detect.rs:23:1
    |
23  | / pub unsafe extern "C" fn rs_ssh_tx_get_protocol(
24  | |     tx: *mut std::os::raw::c_void,
25  | |     buffer: *mut *const u8,
26  | |     buffer_len: *mut u32,
...   |
29  | |     let tx = cast_pointer!(tx, SSHTransaction);
    | |              --------------------------------- in this macro invocation
...   |
52  | |     return 0;
53  | | }
    | |_- because it's nested under this `unsafe` fn
    |
    = note: #[warn(unused_unsafe)] on by default
```

Reply:
Oh, yeah. I think what is needed is a sweeping change to cover this. For now, stick with  your code as-is.